finality-aleph/src/party/mod.rs

```diff
@@ -254,6 +258,342 @@ where
+        let session_manager = Arc::new(MockNodeSessionManager::new());
+        let session_info = MockSessionInfo::new(SESSION_PERIOD);
+
+        let controller = MockController {
+            shared_session_map: shared_map,
+            _sync_state_mock: sync_state.clone(),
+            chain_state_mock: chain_state.clone(),
+            node_session_manager: session_manager.clone(),
+        };
+
+        let params = ConsensusPartyParams {
+            session_authorities: readonly_session_authorities,
+            chain_state,
+            sync_state,
+            backup_saving_path: None,
+            session_manager,
+            session_info,
+            _phantom: Default::default(),
+        };
+
+        (ConsensusParty::new(params), controller)
+    }
+
+    fn simulate_n_blocks(start: u32, n: u32, controller: &MockController) {
+        for i in start..start + n {
+            controller.chain_state_mock.set_best_block(i);
+            controller.chain_state_mock.set_finalized_block(i);
+        }
+    }
+
+    async fn set_authorities_for_sessions(
+        controller: &mut MockController,
+        sessions: Vec<(SessionId, Vec<AuthorityId>)>,
+    ) {
+        for (session, authorities) in sessions {
+            controller
+                .shared_session_map
+                .update(session, SessionAuthorityData::new(authorities, None))
+                .await;
+        }
+    }
+
+    fn assert_session_states(
+        controller: &MockController,
+        validator_started: Vec<SessionId>,
+        early_started: Vec<SessionId>,
+        stopped: Vec<SessionId>,
+        non_validator_started: Vec<SessionId>,
+    ) {
+        assert_eq!(
+            *controller
+                .node_session_manager
+                .validator_session_started
+                .lock()
+                .unwrap(),
+            HashSet::from_iter(validator_started),
+            "`validator_session_started` mismatch"
+        );
+
+        assert_eq!(
+            *controller
+                .node_session_manager
+                .session_early_started
+                .lock()
+                .unwrap(),
+            HashSet::from_iter(early_started),
+            "`session_early_started` mismatch"
+        );
+
+        assert_eq!(
+            *controller
+                .node_session_manager
+                .session_stopped
+                .lock()
+                .unwrap(),
+            HashSet::from_iter(stopped),
+            "`session_stopped` mismatch"
+        );
+
+        assert_eq!(
+            *controller
+                .node_session_manager
+                .nonvalidator_session_started
+                .lock()
+                .unwrap(),
+            HashSet::from_iter(non_validator_started),
+            "`nonvalidator_session_started` mismatch"
+        );
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn party_starts_session_for_node_in_authorities() {
```

Comment:
I've got some problem with this test - it runs full session and actually checks two things:
 - that the provided authorities are indeed in use
 - after the last block of the session is processed, a new session 'begins', i.e. new authority set is used

I would split this into two tests: one running for a part of session (that checks the first condition) and the second one that runs exactly SessionPeriod blocks and checks the second condition

Reply:
it doesn't check the first condition